src/channel_manager.ts

```diff
@@ -0,0 +1,517 @@
+  shouldConsiderArchivedChannels,
+  shouldConsiderPinnedChannels,
+} from './utils';
+
+export type ChannelManagerPagination<SCG extends ExtendableGenerics = DefaultGenerics> = {
+  filters: ChannelFilters<SCG>;
+  hasNext: boolean;
+  isLoading: boolean;
+  isLoadingNext: boolean;
+  options: ChannelOptions;
+  sort: ChannelSort<SCG>;
+  stateOptions: ChannelStateOptions;
+};
+
+export type ChannelManagerState<SCG extends ExtendableGenerics = DefaultGenerics> = {
+  channels: Channel<SCG>[];
+  pagination: ChannelManagerPagination<SCG>;
+  ready: boolean;
+};
+
+export type SetterParameterType<T> = T | ((prevState: T) => T);
+export type ChannelSetterParameterType<SCG extends ExtendableGenerics = DefaultGenerics> = SetterParameterType<
+  ChannelManagerState<SCG>['channels']
+>;
+export type ChannelSetterType<SCG extends ExtendableGenerics = DefaultGenerics> = (
+  arg: ChannelSetterParameterType<SCG>,
+) => void;
+
+// TODO: Figure out a better way to infer a generic handler type here
+export type GenericEventHandlerType<T extends unknown[]> = (
+  ...args: T
+) => void | (() => void) | ((...args: T) => Promise<void>);
+export type EventHandlerType<SCG extends ExtendableGenerics = DefaultGenerics> = GenericEventHandlerType<[Event<SCG>]>;
+export type EventHandlerOverrideType<SCG extends ExtendableGenerics = DefaultGenerics> = GenericEventHandlerType<
+  [ChannelSetterType<SCG>, Event<SCG>]
+>;
+
+export type ChannelManagerEventTypes =
+  | 'notification.added_to_channel'
+  | 'notification.message_new'
+  | 'notification.removed_from_channel'
+  | 'message.new'
+  | 'member.updated'
+  | 'channel.deleted'
+  | 'channel.hidden'
+  | 'channel.truncated'
+  | 'channel.visible';
+
+export type ChannelManagerEventHandlerNames =
+  | 'channelDeletedHandler'
+  | 'channelHiddenHandler'
+  | 'channelTruncatedHandler'
+  | 'channelVisibleHandler'
+  | 'newMessageHandler'
+  | 'memberUpdatedHandler'
+  | 'notificationAddedToChannelHandler'
+  | 'notificationNewMessageHandler'
+  | 'notificationRemovedFromChannelHandler'
+  | 'userPresenceHandler';
+
+export type ChannelManagerEventHandlerOverrides<SCG extends ExtendableGenerics = DefaultGenerics> = Partial<
+  Record<ChannelManagerEventHandlerNames, EventHandlerOverrideType<SCG>>
+>;
+
+const eventToHandlerMapping: { [key in ChannelManagerEventTypes]: ChannelManagerEventHandlerNames } = {
+  'channel.deleted': 'channelDeletedHandler',
+  'channel.hidden': 'channelHiddenHandler',
+  'channel.truncated': 'channelTruncatedHandler',
+  'channel.visible': 'channelVisibleHandler',
+  'message.new': 'newMessageHandler',
+  'member.updated': 'memberUpdatedHandler',
+  'notification.added_to_channel': 'notificationAddedToChannelHandler',
+  'notification.message_new': 'notificationNewMessageHandler',
+  'notification.removed_from_channel': 'notificationRemovedFromChannelHandler',
+};
+
+export type ChannelManagerOptions = {
+  allowNewMessagesFromUnfilteredChannels?: boolean;
+  lockChannelOrder?: boolean;
```

Comment:
Could `skipInitialization` and `offlineMode` go here? Would it make sense?

Reply:
I actually thought a lot about this as well, decided to move them into the reactive state in the end because they seemed `pagination` related. But getting into it further, it does indeed make sense to move 'em out of here as they're mostly related to different initialization behaviour of the state if provided. 

I'll do that in a bit, thanks !